sort the mess with isGB

fieker opened the pull request:
follow up of Tommy's fix earlier, this now adresses this more generally

Also: add simplify(MPolyQuoElem)
          inv, isunit and isunit_with_inverse for those elements
(and a test for inv)

fieker committed:
sort the mess with isGB

follow up of Tommy's fix earlier, this now adresses this more generally

Also: add simplify(MPolyQuoElem)
          inv, isunit and isunit_with_inverse for those elements
(and a test for inv)

thofma commented on src/Rings/MPolyQuo.jl:
```diff
@@ -105,22 +115,48 @@ function quo(R::MPolyRing, I::MPolyIdeal)
-
 one(Q::MPolyQuo) = Q(1)
 
+function isunit_with_inverse(a::MPolyQuoElem)
```
```suggestion
function isinvertible_with_inverse(a::MPolyQuoElem)
```

thofma commented on src/Rings/MPolyQuo.jl:
```diff
@@ -105,22 +115,48 @@ function quo(R::MPolyRing, I::MPolyIdeal)
+
+isunit(a::MPolyQuoElem) = isunit_with_inverse(a)[1]
+function inv(a::MPolyQuoElem)
+  fl, b = isunit_with_inverse(a)
```
```suggestion
  fl, b = isinvertible_with_inverse(a)
```

thofma commented on src/Rings/MPolyQuo.jl:
```diff
@@ -105,22 +115,48 @@ function quo(R::MPolyRing, I::MPolyIdeal)
+  return false, a
+end
+
+isunit(a::MPolyQuoElem) = isunit_with_inverse(a)[1]
```
```suggestion
isunit(a::MPolyQuoElem) = isinvertible_with_inverse(a)[1]

```

thofma reviewed (commented):
We could use `isinvertible_with_inverse`, since we already have it for matrices.

fieker commented:
On Tue, Feb 02, 2021 at 04:29:35AM -0800, thofma wrote:
> @thofma commented on this pull request.
> 
> We could use `isinvertible_with_inverse`, since we already have it for matrices.
> 
will do - Dan just asked also for divisibility testing, so will add as
well
> >  one(Q::MPolyQuo) = Q(1)
>  
> +function isunit_with_inverse(a::MPolyQuoElem)
> 
> ```suggestion
> function isinvertible_with_inverse(a::MPolyQuoElem)
> ```
> 
> > +    J = I.gb.O
> +  else
> +    J = gens(I)
> +  end
> +  J = vcat(J, [a.f])
> +  H, T = groebner_basis_with_transform(ideal(J))
> +  if 1 in H
> +    @assert nrows(T) == 1
> +    return true, Q(T[1, end])
> +  end
> +  return false, a
> +end
> +
> +isunit(a::MPolyQuoElem) = isunit_with_inverse(a)[1]
> +function inv(a::MPolyQuoElem)
> +  fl, b = isunit_with_inverse(a)
> 
> ```suggestion
>   fl, b = isinvertible_with_inverse(a)
> ```
> 
> > +  I = Q.I
> +  if isdefined(I, :gb)
> +    J = I.gb.O
> +  else
> +    J = gens(I)
> +  end
> +  J = vcat(J, [a.f])
> +  H, T = groebner_basis_with_transform(ideal(J))
> +  if 1 in H
> +    @assert nrows(T) == 1
> +    return true, Q(T[1, end])
> +  end
> +  return false, a
> +end
> +
> +isunit(a::MPolyQuoElem) = isunit_with_inverse(a)[1]
> 
> ```suggestion
> isunit(a::MPolyQuoElem) = isinvertible_with_inverse(a)[1]
> 
> ```
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/oscar-system/Oscar.jl/pull/259#pullrequestreview-581290857

fieker committed:
add tests for the graded modules.

Minimal tests...

fieker committed:
add divides for MPOlyQuoElem

and refactor conversion from Singular.Module to the world

fieker committed:
cleanup:

- adjust to the new conversion to/from Singular slpoly
- use the hew conversion to/from Singular.Module

fieker committed:
change to make Julia 1.4 happier

fieker committed:
"fix" test
